### DISCUSSION TITLE
perltidy as a linter ?

### DISCUSSION THREAD

**Discussion by WhoIsSethDaniel** [perltidy/perltidy#123] [Q&A] [answered]
It seems `perltidy` has at least some linting capabilities.  If I use `perltidy --assert-tidy --standard-error-output --outfile=/dev/null` I get output for the first error found. This can be parsed and used by tools that run linters. However, I would like to get *all* errors. Is this possible? I've read through the `perldoc` for `perltidy` but I didn't see anything that would do this. Is this something `perltidy` should be able to do? Or will do eventually?

**Comment by WhoIsSethDaniel** [comment]
I suppose this is an answer: https://github.com/perltidy/perltidy/issues/30.

**Comment by shancock9** [comment, marked as answer]
Yes, this discussion you found is the background for the why perltidy just shows the first difference when an ``--assert-tidy`` signals an error (or not really an error, but rather a formatting difference). For a more complete result, a ``diff`` utility would be required to compactly show the difference between the input and output. It would be difficult to implement this inside of perltidy in a way which worked on all systems that it currently supports (back to perl 5.8).  But it would be fairly simple to write a wrapper program which called the ``Perl::Tidy`` module and then ran a system ``diff`` utility.

The main goal of perltidy is to minimize the work needed to produce readable Perl scripts. It does report any obvious errors that it finds, but that is more of a uesful side effect than the main goal. I wrote the first version about 2000 after realizing that too many of my man hours were being wasted on reformatting a big physics simulation program that I continually messed up with lots of cutting and pasting during debug sessions.

In my own work, after I get a program working, I find it useful to run ``PerlCritic`` to check for problems.  I see it as more closely filling the role of a linter.

**Reply by WhoIsSethDaniel** [reply]
Thank you.

**Reply by oalders** [reply]
Thanks, @shancock9! At $work we use `perlcritic` and `perltidy` as linters. Having a failure on `perltidy` finding a difference allows us to have minimal diffs in code changes, since the style will be enforced across the board. I understand your reasoning, but just wanted to point out that some of us are abusing `perltidy` in ways it was possibly never meant to be used. 😄
